test/fiber/scheduler.rb

```diff
@@ -136,13 +135,51 @@ def wait_any(io, events, duration)
   end
 
   def wait_for_single_fd(fd, events, duration)
-    wait_any(
-      for_fd(fd),
-      events,
-      duration
-    )
+    wait_any(for_fd(fd), events, duration)
   end
 
+  # called from Mutex#lock
+  def wait_mutex(mutex)
+    p [:wait_mutex, mutex]
+    # @waiting_on[Fiber.current] = mutex
+
+    # Guarantee: Fiber.current is a Fiber of current Thread and is a Fiber of this scheduler
+    @waiting_mutex[mutex] << Fiber.current
+    Fiber.yield
+    true
+  end
+
+  # called from Mutex#unlock, possibly from another Fiber on the same thread, or possibly from some other thread
+  def notify_mutex(mutex)
+    p [:notify_mutex, mutex]
+    # p @waiting_mutex
+    q = @waiting_mutex[mutex]
+    fiber = q.shift
+    @waiting_mutex.delete(mutex) if q.empty?
+    raise unless fiber
+
+    if Thread.current == @thread
+      # but what if fiber died in the meanwhile?
+      # Is it possible if same thread? Yes, due to Fiber#raise
+      # And what if it somehow went further in the meanwhile, due to e.g. Fiber#raise, we'll resume another point?
+      if fiber.alive?
+        p [:resume, fiber]
+        fiber.resume
+      end
+    else
+      # scheduler = Thread.current.scheduler
+      # send -> { notify_mutex(mutex) } to scheduler Queue + control pipe write to wake up?
+      # (could be abstracted as #external_notification(&block) ?)
```

Comment:
@ioquatix What do you think of this idea?
I had something similar in Mozart2 and it's super convenient to communicate between threads/Ractors/etc.

Reply:
I don't disagree with something like this.

A way to push work into the scheduler in a thread safe way.

It can be implemented efficiently, and async already supports something very similar to this (but not this).

I've also implemented this in a game engine... it was used for sending asynchronous events, e.g. from the physics engine to the audio subsystem (running on different threads with separate event loops).

The best implementation uses a double buffered event queue.

We might need a lower level primitive than a mutex to implement it correctly, if mutex is fiber aware, it could break semantics if used directly in the scheduler. Or, we need to track blocking/non-blocking and mutex should only change it's behaviour in non-blocking context.